diff --git a/.swp b/.swp
new file mode 100644
index 0000000..60cb313
Binary files /dev/null and b/.swp differ
diff --git a/.vscode/settings.json b/.vscode/settings.json
new file mode 100644
index 0000000..70e34ec
--- /dev/null
+++ b/.vscode/settings.json
@@ -0,0 +1,3 @@
+{
+    "C_Cpp.errorSquiggles": "disabled"
+}
\ No newline at end of file
diff --git a/Makefile b/Makefile
index 212a89d..e3f52f9 100644
--- a/Makefile
+++ b/Makefile
@@ -188,6 +188,11 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_sleep\
+	$U/_pingpong\
+	$U/_primes\
+	$U/_find\
+	$U/_xargs\
 
 
 
diff --git a/time.txt b/time.txt
new file mode 100644
index 0000000..1e8b314
--- /dev/null
+++ b/time.txt
@@ -0,0 +1 @@
+6
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..e5255f4
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,86 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+#include "kernel/fcntl.h"
+
+char* fmtname(char *path){
+    static char buf[DIRSIZ+1];
+    char *p;
+
+    // Find first character after last slash.
+    for(p=path+strlen(path); p >= path && *p != '/'; p--)
+    ;
+    p++;
+
+    // Return blank-padded name.
+    if(strlen(p) >= DIRSIZ)
+    return p;
+    memmove(buf, p, strlen(p));
+    memset(buf+strlen(p), ' ', DIRSIZ-strlen(p));
+    return buf;
+}
+
+void find(char *path, char *target){
+    char buf[512], *p;
+    int fd;
+    struct dirent de;
+    struct stat st;
+
+    if((fd = open(path, O_RDONLY)) < 0){
+        fprintf(2, "find: cannot open %s\n", path);
+        return;
+    }
+
+    if(fstat(fd, &st) < 0){
+        fprintf(2, "find: cannot stat %s\n", path);
+        close(fd);
+        return;
+    }
+
+    switch(st.type){
+        case T_DEVICE:
+
+        case T_FILE:
+
+        case T_DIR:
+            if(strlen(path) + 1 + DIRSIZ + 1 > sizeof buf){
+                printf("find: path too long\n");
+                break;
+            }
+            strcpy(buf, path);
+            p = buf + strlen(buf);
+            *p++ = '/';
+            while(read(fd, &de, sizeof(de)) == sizeof(de)){
+                if(de.inum == 0)
+                    continue;
+                memmove(p, de.name, DIRSIZ);
+                p[DIRSIZ] = 0;
+                if(stat(buf, &st) < 0){
+                    printf("find: cannot stat %s\n", buf);
+                    continue;
+                }
+                
+                if(!strcmp(de.name, ".") || !strcmp(de.name, "..")) continue;
+
+                if(st.type == T_DIR){
+                    find(buf, target);
+                }else if(st.type == T_FILE){
+                    if(!strcmp(target, de.name)) printf("%s\n", buf);
+                }
+            }
+            break;
+    }
+    close(fd);
+}
+
+int main(int argc, char *argv[]){
+    if(argc != 3){
+        printf("error\n");
+        exit(0);
+    }
+    
+    find(argv[1], argv[2]);
+
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/pingpong.c b/user/pingpong.c
new file mode 100644
index 0000000..d2fb76e
--- /dev/null
+++ b/user/pingpong.c
@@ -0,0 +1,32 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[]){
+    int fds1[2], fds2[2], pid;
+    char buf[512];
+
+    pipe(fds1);
+    pipe(fds2);
+
+    pid = fork();
+
+    if(pid){
+        write(fds1[0], "Don't forget 3.Oct\n", 19);
+    }else{
+        read(fds1[1], buf, sizeof(buf));
+        printf("%d: received ping\n", getpid());
+        write(fds2[0], "I won't\n", 8);
+        exit(0);
+    }
+
+    
+    if(pid){
+        wait(0);
+        read(fds2[1], buf, sizeof(buf));
+        printf("%d: received pong\n", getpid());
+    }
+    
+
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/primes.c b/user/primes.c
new file mode 100644
index 0000000..372ee72
--- /dev/null
+++ b/user/primes.c
@@ -0,0 +1,40 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[]){
+    int fds[2], pid, n, p;
+    
+    pipe(fds);
+
+    pid = fork();
+
+    if(!pid){
+        close(fds[0]);
+        for(int i = 2; i <= 35; i++) write(fds[1], &i, sizeof(int));
+        exit(0);
+    }else{
+        wait(0);
+        close(fds[1]);
+
+        while(read(fds[0], &p, sizeof(int)) > 0){
+            printf("prime %d\n", p);
+
+            int fdsn[2];
+            pipe(fdsn);
+
+            pid = fork();
+            if(!pid){
+                while(read(fds[0], &n, sizeof(int)) > 0) if(n % p) write(fdsn[1], &n, sizeof(int));
+                exit(0);
+            }else{
+                wait(0);
+                close(fds[0]);
+                fds[0] = fdsn[0];
+                close(fdsn[1]);
+            }
+        }
+    }
+
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..eb84f82
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,15 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[]){
+    if(argc != 2){
+        printf("error");
+        exit(0);
+    }
+
+    char* time = argv[1];
+    sleep(atoi(time));
+    
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/user.h b/user/user.h
index 4d398d5..abc4837 100644
--- a/user/user.h
+++ b/user/user.h
@@ -1,5 +1,6 @@
 struct stat;
 
+
 // system calls
 int fork(void);
 int exit(int) __attribute__((noreturn));
diff --git a/user/xargs.c b/user/xargs.c
new file mode 100644
index 0000000..f3f5c07
--- /dev/null
+++ b/user/xargs.c
@@ -0,0 +1,47 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/param.h"
+
+#define true 1
+
+int main(int argc, char *argv[]) {
+    char *args[MAXARG];
+    char buf[128], ch;
+    int pid, curr;
+    
+    if (argc < 2) {
+        printf("error");
+        exit(0);
+    }
+
+    for (int i = 1; i < argc; i++) args[i - 1] = argv[i];
+    args[argc] = 0;
+    
+    while (true) {
+        curr = 0;
+        
+        while (read(0, &ch, sizeof(char)) == 1) {
+            if (ch == '\n') break;
+            buf[curr] = ch;
+            curr++;
+        }
+        
+        if (!curr) break;
+
+        buf[curr] = '\0';
+
+        pid = fork();
+        
+        if (!pid) {
+            args[argc - 1] = buf;
+            exec(args[0], args);
+            printf("exec failed!\n");
+            exit(1);
+        } else {
+            wait(0);
+        }
+    }
+
+    exit(0);
+}
\ No newline at end of file
